apollo-backend/src/test/java/io/logz/apollo/SlaveTest.java

```diff
@@ -0,0 +1,79 @@
+import io.logz.apollo.helpers.ModelsGenerator;
+import io.logz.apollo.helpers.StandaloneApollo;
+import io.logz.apollo.kubernetes.KubernetesMonitor;
+import io.logz.apollo.models.Environment;
+import io.logz.apollo.services.SlaveService;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.script.ScriptException;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class SlaveTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(SlaveTest.class);
+    private final StandaloneApollo standaloneApollo;
+
+    public SlaveTest() throws ScriptException, IOException, SQLException {
+        this.standaloneApollo = StandaloneApollo.getOrCreateServer();
+    }
+
+    @Test
+    public void testSlaveOwnership() throws ApolloClientException {
+        ApolloTestClient apolloTestClient = Common.signupAndLogin();
+        EnvironmentDao environmentDao = standaloneApollo.getInstance(EnvironmentDao.class);
+
+        Environment slaveEnvironment1 = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+        Environment slaveEnvironment2 = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+        Environment masterEnvironment1 = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+        Environment masterEnvironment2 = ModelsGenerator.createAndSubmitEnvironment(apolloTestClient);
+
+        List<Integer> allEnvironmentIds =  apolloTestClient.getAllEnvironments().stream().map(Environment::getId).collect(Collectors.toList());
+        List<Integer> masterScopedEnvironments = standaloneApollo.getInstance(SlaveService.class).getScopedEnvironments();
+        assertThat(getCsvFromList(allEnvironmentIds)).isEqualTo(getCsvFromList(masterScopedEnvironments));
+
+        ApolloApplication slave = standaloneApollo.createAndStartSlave(Arrays.asList(slaveEnvironment1.getId(), slaveEnvironment2.getId()));
+        waitUntilSlaveStarted(slave);
+
+        masterScopedEnvironments = standaloneApollo.getInstance(SlaveService.class).getScopedEnvironments();
+        assertThat(masterScopedEnvironments).contains(masterEnvironment1.getId(), masterEnvironment2.getId());
+        assertThat(masterScopedEnvironments).doesNotContain(slaveEnvironment1.getId(), slaveEnvironment2.getId());
+
+        List<Integer> slaveScopedEnvironments = slave.getInjector().getInstance(SlaveService.class).getScopedEnvironments();
+        assertThat(slaveScopedEnvironments).containsOnly(slaveEnvironment1.getId(), slaveEnvironment2.getId());
+    }
+
+    private void waitUntilSlaveStarted(ApolloApplication apolloApplication) {
+        SlaveService slaveService = apolloApplication.getInjector().getInstance(SlaveService.class);
+        int retry = 10;
+        for(int i = 0; i <= retry; i++) {
```

Comment:
Add dependency for failsafe if missing and use it instead

Reply:
I don't think it's necessary to add this dependency only for this use

Reply:
It is, especially when it's a test dependency.
There's no need to implement something which already exists